### DISCUSSION TITLE
[JAVA] pre-defined query fails to detect CWE-502.

### DISCUSSION THREAD

**Discussion by jiezhuzzz** [github/codeql#16081] [Q&A]
I'm running pre-defined [query](https://github.com/github/codeql/blob/main/java/ql/src/Security/CWE/CWE-502/UnsafeDeserialization.ql) to detect a identified CWE-502 vulnerability jfinal/jfinal#184. The [root cause](https://github.com/jfinal/jfinal/blob/076e76a4941c989d7f535035295aa3d055123131/src/main/java/com/jfinal/plugin/redis/serializer/JdkSerializer.java#L75) of this issue is clear, as it use the `readObject` function without any extra operations. But I got a blank result after I ran the pre-defined query. I took a look at the implementation, and I tried this predicate which worked well. https://github.com/github/codeql/blob/5b1cae5fc20c7f476a413451e4cc02b5f17bc688/java/ql/lib/semmle/code/java/security/UnsafeDeserializationQuery.qll#L148-L235 Any hints about why the pre-defined query cannot detect this issue? Thanks a lot!

**Comment by smowton** [comment]
I think the answer is that the unsafe-deserialization query finds that vulnerable use-site alright, but all the relevant **sources** of untrusted information within JFinal are in the core module (e.g., reading HTTP header parameters), while the unsafe code could only be reached via `Redis.get` and similar functions, but JFinal doesn't itself call those (presumably, user programs built on JFinal would join the two things together).

Another way to say that is, we don't know _from the code of JFinal alone_ what your threat model is: is the stuff stored in Redis safe (you sanitized it on its way into the database), in which case it's fine to deserialize it to an arbitrary object, or is it potentially dangerous?

In general something like this would be handled by customising the query to regard `jedis.get` as a source of dangerous (untrusted) data, and so instructing CodeQL to check for that data flowing to a dangerous call such as `readObject`.

**Reply by jiezhuzzz** [reply]
Hi Chris, thanks for your reply! If I understand correctly,  CodeQL cannot detect if a source in project A and a sink in project B, because 1. it does not constitute a complete dataflow 2. it is unknown whether sanitization is performed in the middle. 

And I found the root cause can be detected as `UnsafeDeserializationSink`, but since the source is not in the JFinal. We cannot detect the whole dataflow.

Thanks again for your help!
